### ISSUE TITLE
Vue | Navigator issue: @prepush event.cancel() should not add the cancelled page to the PageStack

### ISSUE THREAD

**Issue by subdigit** [OnsenUI/OnsenUI#2724]
__Environment__

  
```
[Core]
  onsenui latest
  
[Framework]
  vue 2.4.1

[Framework binding]
 vue-onsenui latest

[Platform]
  Desktop - OS X 10.14.6 64-bit

[Browser]
  Desktop - Chrome 82.0.4051.0 on OS X 10.14.6 64-bit
```

__Encountered problem__
When cancelling an event based on @prepush on the v-ons-navigator, it successfully cancels the push event, but the page is still added to the pageStack.  This bloats the pageStack and it no longer represents the actual navigation stack.

Ideally, when the @prepush is event.cancel(), the page should not be added to the pageStack to represent that it was cancelled so it never should have been considered as successfully navigated to.

For stack management that simply pushes pages onto the stack, this bloats the stack with unnavigated to pages.  For those that keep a clean, non repeating stack, it will be considered to already be on that cancelled page even if technically the user never navigated to it, so subsequent attempts to push that same page onto the stack will make the system think it's already on that page.

__How to reproduce__
From the demo, click "Push Page 2" then back, same with "Push Page 3".  Those work normally.  Click "Cancel next page" which will allow the @prepush code to do an event.cancel().  Now do a "Push Page 2" which properly does nothing.

In the JS console, there should be a note showing the current page stack count (it should show "2" for Page 1 and Page 2.

Click "Push Page 2" over and over.  Navigation is still disabled so nothing happens, but the console will be reporting a higher and higher pageStack count.

  
  [__Demo link__](https://onsen.io/playground/?issue)

  

- __HTML__

```html
<template id="main">
  <v-ons-navigator swipeable
    :page-stack="pageStack"
    @push-page="setPage($event)"
    @cancel-page="cancelPage($event)"
    @prepush="checkStack"
  ></v-ons-navigator>
</template>

<template id="page1">
  <v-ons-page>
    <v-ons-toolbar>
      <div class="center">Page 1</div>
    </v-ons-toolbar>
    <p style="text-align: center">
      This is the first page
      <br /><v-ons-button @click="push">Push Page 2</v-ons-button>
      <br /><v-ons-button @click="push3">Push Page 3</v-ons-button>
      <br /><v-ons-button @click="cancel">Cancel next page</v-ons-button>
      <br /><v-ons-button @click="uncancel">Uncancel next page</v-ons-button>
    </p>
  </v-ons-page>
</template>

<template id="page2">
  <v-ons-page>
    <v-ons-toolbar>
      <div class="left">
        <v-ons-back-button>Back</v-ons-back-button>
      </div>
      <div class="center">Page 2</div>
    </v-ons-toolbar>
    <p style="text-align: center">This is the second page</p>
  </v-ons-page>
</template>

<template id="page3">
  <v-ons-page>
    <v-ons-toolbar>
      <div class="left">
        <v-ons-back-button>Back</v-ons-back-button>
      </div>
      <div class="center">Page 3</div>
    </v-ons-toolbar>
    <p style="text-align: center">This is the third page</p>
  </v-ons-page>
</template>

<div id="app"></div>
```

- __JS__

```javascript
const page3 = {
  key: 'page3',
  template: '#page3'
};

const page2 = {
  key: 'page2',
  template: '#page2'
};

const page1 = {
  key: 'page1',
  template: '#page1',
  methods: {
    push() {
      this.$emit('push-page', page2);
    },
    push3() {
      this.$emit('push-page', page3);
    },
    cancel() {
      this.$emit('cancel-page', true);
    },
    uncancel() {
      this.$emit('cancel-page', false);
    }
  }
};

new Vue({
  el: '#app',
  template: '#main',
  data() {
    return {
      pageStack: [page1],
      cancelNavigation: false
    };
  },
  methods:{
    setPage(event){
      this.pageStack.push(event);
    },
    cancelPage(event){
      this.cancelNavigation = event;
    },
    checkStack(event){
      console.log(this.pageStack);
      if(!this.cancelNavigation) console.log("loading new page");
      else {
        event.cancel();

        console.log("cancelled the event, but the page still got added to the stack");
        console.log("If I try to pop the page off the stack, the reactive nature will cause the navigation to happen again.");
        console.log("PageStack: ");
        console.log(this.pageStack);
//        this.pageStack.pop();
      }
    }
  }
});
```

**Comment by emccorson** [comment]
@subdigit Thank you for the detailed bug report!

I was able to replicate the bug, and I've confirmed that it is a bug in the Vue bindings and not in the Onsen UI core.

The problem is that a page push is initiated by pushing a new element to the Vue page stack with `this.pageStack.push`. The watcher on `pageStack` is then triggered and calls the native `this.$el._pushPage`. However, if the native push is canceled, we have to remove the element that was pushed to the Vue page stack. This has to be done without Vue thinking `pageStack` has been changed again because it will want to show the pop page animation.

Working on a fix now...

**Comment by subdigit** [comment]
Question, with the fix in 7f9bb2f, does this mean that if I was independently monitoring the `pageStack` (like in a watch) I might still also get reactively notified about the `.pop()` because of the cancelled event?

I was having that problem in my code, and I had to resort to non reactively removing the most recent item in the pageStack by exploiting one of the Vue's Array reactivity caveats:             `this.pageStack.length -= 1;`

Haven't sat down with the checked in code to see it if it does indeed suffer the same problem I ran into with using `.pop()`.

P.S.  Thanks for the fix!
